sklearn/metrics/cluster/supervised.py

```diff
@@ -849,24 +879,50 @@ def fowlkes_mallows_score(labels_true, labels_pred, sparse=False):
     .. [2] `Wikipedia entry for the Fowlkes-Mallows Index
            <https://en.wikipedia.org/wiki/Fowlkes-Mallows_index>`_
     """
-    labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
-    n_samples, = labels_true.shape
+    
+    if labels_pred is None:
+        # labels_true is actually contingency
+        contingency = check_array(labels_true,
+                                  accept_sparse= True,
+                                  dtype=[int, np.int32, np.int64])
+        if not sp.issparse(contingency):
+            contingency = sp.csr_matrix(contingency)
+        n_samples = contingency.sum()
+
 
-    c = contingency_matrix(labels_true, labels_pred, sparse=True)
-    tk = np.dot(c.data, c.data) - n_samples
-    pk = np.sum(np.asarray(c.sum(axis=0)).ravel() ** 2) - n_samples
-    qk = np.sum(np.asarray(c.sum(axis=1)).ravel() ** 2) - n_samples
+    else:
+        labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
+        n_samples, = labels_true.shape
+
+    contingency = contingency_matrix(labels_true, labels_pred, sparse=True)
+    tk = np.dot(contingency.data, contingency.data) - n_samples
+    pk = np.sum(np.asarray(contingency.sum(axis=0)).ravel() ** 2) - n_samples
+    qk = np.sum(np.asarray(contingency.sum(axis=1)).ravel() ** 2) - n_samples
     return tk / np.sqrt(pk * qk) if tk != 0. else 0.
 
 
-def entropy(labels):
+def entropy(data):
     """Calculates the entropy for a labeling."""
-    if len(labels) == 0:
-        return 1.0
-    label_idx = np.unique(labels, return_inverse=True)[1]
-    pi = bincount(label_idx).astype(np.float64)
-    pi = pi[pi > 0]
-    pi_sum = np.sum(pi)
-    # log(a / b) should be calculated as log(a) - log(b) for
-    # possible loss of precision
-    return -np.sum((pi / pi_sum) * (np.log(pi) - log(pi_sum)))
+    if data.ndim == 2:
```

Comment:
This one seems to be the culprit. You have assumed the data to be `ndarray` which it might not be. Please correct this.